compiler/next/lib/resolution/scope-queries.cpp

```diff
@@ -0,0 +1,529 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chpl/resolution/scope-queries.h"
+
+#include "chpl/parsing/parsing-queries.h"
+#include "chpl/queries/ErrorMessage.h"
+#include "chpl/queries/query-impl.h"
+#include "chpl/uast/all-uast.h"
+
+#include <cstdio>
+#include <set>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace chpl {
+namespace resolution {
+
+
+using namespace uast;
+using namespace types;
+
+struct GatherDecls {
+  DeclMap declared;
+  bool containsUseImport = false;
+
+  GatherDecls() { }
+
+  // Add NamedDecls to the map
+  bool enter(const NamedDecl* d) {
+    UniqueString name = d->name();
+    auto search = declared.find(name);
+    if (search == declared.end()) {
+      // add a new entry containing just the one ID
+      declared.emplace_hint(search,
+                            name,
+                            OwnedIdsWithName(d->id()));
+    } else {
+      // found an entry, so add to it
+      OwnedIdsWithName& val = search->second;
+      val.appendId(d->id());
+    }
+    return false;
+  }
+  void exit(const NamedDecl* d) { }
+
+  // Traverse into TupleDecl and MultiDecl looking for NamedDecls
+  bool enter(const TupleDecl* d) {
+    return true;
```

Comment:
Since it's a traversal visitor, and the TupleDecl/MultiDecl contains NamedDecls, we will handle the NamedDecls when visiting the children. Returning true means we visit the children.